File: SILDMS/SILDMS.Web.UI/Scripts/App/Controller/DocTypeController.js
//// Controller for Document Type page
//app.controller("docTypeCtrl", ["$scope", "$http", function ($scope, $http) {
//    // Model for Document Type
//    $scope.docType = {
//        DocTypeID: "", OwnerLevel: { OwnerLevelID: "", LevelName: "" }, Owner: { OwnerID: "", OwnerName: "" },
//        DocCategory: { DocCategoryID: "", DocCategoryName: "" }, DocTypeSL: "", UDDocTypeCode: "", DocTypeName: "", DocPreservationPolicy: "",
//        DocPhysicalLocation: "", Status: ""
//    };
//    // Model for the data to saved/updated
//    $scope.saveModel = {};
//    $scope.itemsByPage = 10;
//    $scope.loading = true;
//    // Get all owner levels
//    $http.get("/DocScanningModule/DocType/GetOwnerLevel?id=")
//        .success(function (response) { $scope.ownerLevels = response.result; $scope.loading = false; })
//        .error(function () { });
//    // Get all owners for the selected owner level on owner level dropdown change
//    $scope.changeOwnerLevel = function () {
//        $http.get("/DocScanningModule/DocType/GetOwners?id=" + $scope.docType.OwnerLevel.OwnerLevelID)
//            .success(function (response) { $scope.owners = response.result; $scope.docTypes = []; $scope.docCategories = []; $scope.loading = true; })
//            .error(function () { });
//    }
//    // Get all document categories for the selected owner on owner drop down change
//    $scope.changeOwner = function () {
//        $http.get("/DocScanningModule/DocType/GetDocCategoryForOwner?id=" + $scope.docType.Owner.OwnerID)
//            .success(function (response) { $scope.docCategories = response.result; $scope.docTypes = []; })
//            .error(function () { });
//    };
//    // Get all document types for the selected document category on documnet category drop down change
//    $scope.changeDocCategory = function () {
//        $http.get("/DocScanningModule/DocType/GetDocTypeForCategory?id=" + $scope.docType.DocCategory.DocCategoryID)
//            .success(function (response) { $scope.docTypes = response.result; $scope.loading = false; })
//            .error(function () { });
//    };
//    // Add new Document Type
//    $scope.toggleAdd = function (row) {
//        $scope.docType.DocTypeID = ""; $scope.docType.DocTypeSL = ""; $scope.docType.UDDocTypeCode = ""; $scope.docType.DocTypeName = "";
//        $scope.docType.DocPreservationPolicy = ""; $scope.docType.DocPhysicalLocation = ""; $scope.docType.Status = ""; $('#addModal').modal('show');
//    };
//    // Edit Document Type
//    $scope.toggleEdit = function (row) {
//        $scope.docType.DocTypeID = row.DocTypeID; $scope.docType.DocTypeSL = row.DocTypeSL; $scope.docType.UDDocTypeCode = row.UDDocTypeCode;
//        $scope.docType.DocTypeName = row.DocTypeName; $scope.docType.DocPreservationPolicy = row.DocPreservationPolicy; $scope.docType.DocPhysicalLocation = row.DocPhysicalLocation;
//        $scope.docType.Status = (row.Status).toString(); $("#addModal").modal("show");
//    };
//    // Convert current model to save model
//    $scope.convertModel = function () {
//        $scope.saveModel.DocTypeID = $scope.docType.DocTypeID; $scope.saveModel.OwnerID = $scope.docType.Owner.OwnerID;
//        $scope.saveModel.DocCategoryID = $scope.docType.DocCategory.DocCategoryID; $scope.saveModel.DocTypeSL = $scope.docType.DocTypeSL;
//        $scope.saveModel.UDDocTypeCode = $scope.docType.UDDocTypeCode; $scope.saveModel.DocTypeName = $scope.docType.DocTypeName;
//        $scope.saveModel.DocPreservationPolicy = $scope.docType.DocPreservationPolicy; $scope.saveModel.DocPhysicalLocation = $scope.docType.DocPhysicalLocation;
//        $scope.saveModel.Status = $scope.docType.Status.toString();
//    };
//    // Save/Update Document Type
//    $scope.Save = function () {
//        $scope.loading = true; $scope.convertModel();
//        if ($scope.docType.DocTypeID) {
//            // Update
//            $http.post("/DocScanningModule/DocType/EditDocType/", JSON.stringify(convArrToObj($scope.saveModel))).success(function (data) {
//                if (data._respStatus.Status === "1") {
//                    $scope.loading = false; scope.changeDocCategory(); ("#addModal").modal("hide");
//                    toastr.success(data.Message);
//                } else { $scope.loading = false; toastr.error(data.Message); }
//            }).error(function (data) { $scope.loading = false; toastr.error("Update Failed."); });
//        } else {
//            // Save
//            $http.post("/DocScanningModule/DocType/AddDocType/", JSON.stringify(convArrToObj($scope.saveModel))).success(function (data) {
//                if (data._respStatus.Status === "1") {
//                    $scope.loading = false; $scope.changeDocCategory(); $("#addModal").modal("hide");
//                    toastr.success(data.Message);
//                } else { $scope.loading = false; toastr.error(data.Message); }
//            })
//                .error(function (data) { $scope.loading = false; toastr.error("Save Failed."); });
//        }
//    }
//    // Refresh Documnet Type Table
//    $scope.toggleRefreshTable = function () { $scope.loading = true; $scope.changeDocCategory(); };
//}]);

File: SILDMS/SILDMS.Web.UI/Areas/DocScanningModule/Controllers/DocTypeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Web;
using System.Web.Mvc;
using SILDMS.Model.DocScanningModule;
using SILDMS.Service.DocumentCategory;
using SILDMS.Service.DocumentType;
using SILDMS.Service.Owner;
using SILDMS.Service.OwnerLevel;
using SILDMS.Utillity;
using SILDMS.Utillity.Localization;
using System.Web.SessionState;
using SILDMS.Model.SecurityModule;
using SILDMS.Service.UserLevel;
using SILDMS.Web.UI.Areas.SecurityModule;
using SILDMS.Web.UI.Areas.SecurityModule.Models;

namespace SILDMS.Web.UI.Areas.DocScanningModule.Controllers
{

    public class DocTypeController : Controller
    {

        #region Fields

        readonly IOwnerService _ownerService;
        private readonly IOwnerLevelService _ownerLevelService;
        private readonly IDocCategoryService _docCategoryService;
        private readonly IDocTypeService _docTypeService;
        private readonly IUserLevelService _userLevelService;
        private readonly ILocalizationService _localizationService;
        private ValidationResult _respStatus = new ValidationResult();
        private string _outStatus = string.Empty;
        private string _action = string.Empty;
        private readonly string UserID = string.Empty;
        #endregion

        #region Constructor

        public DocTypeController(IDocTypeService docTypeService, ILocalizationService localizationService, 
            IOwnerService ownerService, IOwnerLevelService ownerLevelService, IDocCategoryService docCategoryService, IUserLevelService userLevelService)
        {
            _docTypeService = docTypeService;
            _localizationService = localizationService;
            _ownerService = ownerService;
            _ownerLevelService = ownerLevelService;
            _docCategoryService = docCategoryService;
            _userLevelService = userLevelService;
            UserID = SILAuthorization.GetUserID();
        }

        #endregion

        // GET: DocScanningModule/DocType
        [SILAuthorize]
        public ActionResult Index()
        {
            return View();
        }

        #region Document Type Manipulation

        // Add New Document Type
        [Authorize]
        [HttpPost]
        [SILLogAttribute]
        public async Task<dynamic> AddDocType(DSM_DocType objDocType)
        {
            if (ModelState.IsValid)
            {
                _action = "add";
                objDocType.SetBy = UserID;
                objDocType.ModifiedBy = objDocType.SetBy;
                _respStatus = await Task.Run(() => _docTypeService.ManipulateDocType(objDocType, _action, out _outStatus));
                // Error handling.   
                return Json(new { Message = _respStatus.Message, _respStatus }, JsonRequestBehavior.AllowGet);
            }
            else
            {
                _respStatus = new ValidationResult("E404", _localizationService.GetResource("E404"));
            }
            return Json(new { Message = _respStatus.Message, _respStatus }, JsonRequestBehavior.AllowGet);
        }

        // Edit Document Type
        [Authorize]
        [HttpPost]
        [SILLogAttribute]
        public async Task<dynamic> EditDocType(DSM_DocType objDocType)
        {
            if (ModelState.IsValid)
            {
                _action = "edit";
                objDocType.SetBy = UserID;
                objDocType.ModifiedBy = objDocType.SetBy;
                _respStatus = await Task.Run(() => _docTypeService.ManipulateDocType(objDocType, _action, out _outStatus));
                return Json(new { Message = _respStatus.Message, _respStatus }, JsonRequestBehavior.AllowGet);
            }
            else
            {
                _respStatus = new ValidationResult("E404", _localizationService.GetResource("E404"));
            }
            return Json(new { Message = _respStatus.Message, _respStatus }, JsonRequestBehavior.AllowGet);
        }

        // Delete Document Type
        [Authorize]
        [HttpPost]
        [SILLogAttribute]
        public async Task<dynamic> DeleteDocType(DSM_DocType objDocType)
        {
            _action = "delete";
            _respStatus = await Task.Run(() => _docTypeService.ManipulateDocType(objDocType, _action, out _outStatus));
            return Json(new { Message = _respStatus.Message, _respStatus }, JsonRequestBehavior.AllowGet);
        }

        #endregion

        #region Get data for view

        [HttpGet]
        public async Task<dynamic> GetOwnerLevel(string id)
        {
            var ownerLevel = new List<DSM_OwnerLevel>();
            await Task.Run(() => _ownerLevelService.GetOwnerLevel(id, UserID, out ownerLevel));
            var result = ownerLevel.Where(ob => ob.Status == 1).Select(x => new
            {
                x.OwnerLevelID,
                x.LevelName
            }).OrderByDescending(ob => ob.LevelName);
            return Json(new { result, Msg = "" }, JsonRequestBehavior.AllowGet);
        }

        // Get Owner(s)
        [HttpGet]
        public async Task<dynamic> GetOwners(string id)
        {
            var owner = new List<DSM_Owner>();
            await Task.Run(() => _ownerService.GetAllOwners("", UserID, out owner));
            var result = owner.Where(ob => ob.OwnerLevelID == id && ob.Status == 1).Select(x => new
            {
                x.OwnerID,
                x.OwnerName
            }).OrderByDescending(ob => ob.OwnerID);
            return Json(new { result, Msg = "" }, JsonRequestBehavior.AllowGet);
        }

        public async Task<dynamic> GetDocCategoryForOwner(string id)
        {
            var docCategories = new List<DSM_DocCategory>();
            await Task.Run(() => _docCategoryService.GetDocCategories("", UserID, out docCategories));
            var result = docCategories.Where(ob => ob.OwnerID == id && ob.Status == 1).Select(x => new
            {
                x.DocCategoryID,
                x.DocCategoryName
            }).OrderByDescending(ob => ob.DocCategoryID);
            return Json(new { result, Msg = "" }, JsonRequestBehavior.AllowGet);
        }

        public async Task<dynamic> GetDocTypeForCategory(string id)
        {
            var docTypes = new List<DSM_DocType>();
            await Task.Run(() => _docTypeService.GetDocTypes("", UserID, out docTypes));
            var result = docTypes.Where(ob => ob.DocCategoryID == id).Select(x => new
            {
                x.DocTypeID,
                x.DocTypeSL,
                x.UDDocTypeCode,
                x.DocTypeName,
                x.DocPreservationPolicy,
                x.DocPhysicalLocation,
                x.DocCategoryID,
                x.OwnerID,
                x.Status,
                x.DocClassification,
                x.ClassificationLevel
                
            }).OrderByDescending(ob => ob.DocCategoryID);
            return Json(new { result, Msg = "" }, JsonRequestBehavior.AllowGet);
        }

        [HttpGet]
        public async Task<dynamic> GetDocClassification(int? userLevel, string levelType)
        {
            var userLevels = new List<SEC_UserLevel>();
            await Task.Run(() => _userLevelService.GetUserLevels(userLevel, _action, levelType, out userLevels));
            var result = userLevels.Select(ob => new
            {
                ob.ID,
                DocClassificationName = ob.UserLevelName,
             
            }).OrderBy(ob => ob.ID);
            return Json(new { result, Msg = "" }, JsonRequestBehavior.AllowGet);
        }

        [HttpGet]
        public async Task<dynamic> GetClassificationLevel(int? userLevel, string levelType)
        {
            var userLevels = new List<SEC_UserLevel>();
            await Task.Run(() => _userLevelService.GetUserLevels(userLevel, _action, levelType, out userLevels));
            var result = userLevels.Select(ob => new
            {
                ob.ID,
                ClassificationLevelName = ob.UserLevelName,

            }).OrderBy(ob => ob.ID);
            return Json(new { result, Msg = "" }, JsonRequestBehavior.AllowGet);
        }
 

        #endregion

    }
}